packages/headless/src/controllers/commerce/core/breadcrumb-manager/headless-core-breadcrumb-manager.ts

```diff
@@ -0,0 +1,219 @@
+ */
+interface BreadcrumbManagerState {
+  /**
+   * The list of facet breadcrumbs.
+   */
+  facetBreadcrumbs: Breadcrumb<AnyFacetValueResponse>[];
+
+  /**
+   * Returns `true` if there are any available breadcrumbs (i.e., if there are any active facet values), and `false` if not.
+   */
+  hasBreadcrumbs: boolean;
+}
+
+/**
+ * The `BreadcrumbManager` controller manages the breadcrumbs.
+ */
+export type BreadcrumbManager = Omit<CoreBreadcrumbManager, 'state'> & {
+  /**
+   * The state of the `BreadcrumbManager` controller.
+   */
+  state: BreadcrumbManagerState;
+};
+
+interface ActionCreators {
+  toggleSelectActionCreator: ToggleActionCreator;
+  toggleExcludeActionCreator?: ToggleActionCreator;
+}
+
+const facetTypeWithoutExcludeAction: FacetType = 'hierarchical';
+
+const actions: Record<FacetType, ActionCreators> = {
+  regular: {
+    toggleSelectActionCreator: toggleSelectFacetValue,
+    toggleExcludeActionCreator: toggleExcludeFacetValue,
+  },
+  numericalRange: {
+    toggleSelectActionCreator: toggleSelectNumericFacetValue,
+    toggleExcludeActionCreator: toggleExcludeNumericFacetValue,
+  },
+  dateRange: {
+    toggleSelectActionCreator: toggleSelectDateFacetValue,
+    toggleExcludeActionCreator: toggleExcludeDateFacetValue,
+  },
+  [facetTypeWithoutExcludeAction]: {
+    toggleSelectActionCreator: toggleSelectCategoryFacetValue,
+  },
+};
+
+/**
+ * @internal
+ * Creates a `BreadcrumbManager` controller instance.
+ *
+ * @param engine - The headless commerce engine.
+ * @param options - The `BreadcrumbManager` options used internally.
+ * @returns A `BreadcrumbManager` controller instance.
+ **/
+export function buildCoreBreadcrumbManager(
+  engine: CommerceEngine,
+  options: CoreBreadcrumbManagerOptions
+): BreadcrumbManager {
+  if (!loadCommerceBreadcrumbManagerReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+
+  const createBreadcrumb = (facetId: string) => {
+    const facet = options.facetResponseSelector(engine.state, facetId);
+
+    if (!facet) {
+      return {
+        facetId,
+        field: '',
+        type: 'regular',
+        values: [],
+      };
+    }
+
+    return {
+      facetId,
+      field: facet.field,
+      type: facet.type,
+      values: facet.values
+        .filter((value) => value.state !== 'idle')
+        .map((selection) => ({
+          value: selection,
+          deselect: () => {
+            if (selection.state === 'selected') {
+              // TODO: Do we need to handle freezing values?
+              dispatch(
+                actions[facet.type].toggleSelectActionCreator({
+                  facetId,
+                  selection,
+                })
+              );
+              dispatch(options.fetchResultsActionCreator());
+            } else if (
+              selection.state === 'excluded' &&
+              facet.type !== facetTypeWithoutExcludeAction
```

Comment:
This makes me think we should do a refactoring to extract the "real" core commerce facet.

The way I see it, we have "flat" facets (everything except category), and category facets. There's an overlap between flat and category facets (the "real" core), but it's not as wide as we initially thought.